### PR TITLE
Move Contact Rollups writes to primary database

### PR DESCRIPTION
**Pull request by sureshc**
https://codedotorg.atlassian.net/browse/INF-129

Previously, Contact Rollups carried out bulk inserts/updates into the `contact_rollups_daily` table and the bulk reads from sources tables in the reporting replica to ensure that performance of the primary database was not impacted.  If/when we migrate the database to Aurora, writes to the replica will not be permitted (MySQL permits disabling the default read-only setting on the replica).

Modify all the bulk insert/update queries (and their embedded select statements) to read from and write to the contact_rollups_daily table in the primary database.

### CODE REVIEW

**Comment by sureshc** [comment]
I believe this change will not impact runtime performance of the production database because MySQL will use `READ COMMITTED` isolation level on the SELECT statements embedded in the INSERT/UPDATE statements:

https://dev.mysql.com/doc/refman/5.7/en/innodb-consistent-read.html

> The type of read varies for selects in clauses like INSERT INTO ... SELECT, UPDATE ... (SELECT), and CREATE TABLE ... SELECT that do not specify FOR UPDATE or LOCK IN SHARE MODE:
> 
> By default, InnoDB uses stronger locks and the SELECT part acts like READ COMMITTED, where each consistent read, even within the same transaction, sets and reads its own fresh snapshot.
> 
> To use a consistent read in such cases, enable the innodb_locks_unsafe_for_binlog option and set the isolation level of the transaction to READ UNCOMMITTED, READ COMMITTED, or REPEATABLE READ (that is, anything other than SERIALIZABLE). In this case, no locks are set on rows read from the selected table.

**Comment by sureshc** [comment]
If my understanding of MySQL transaction isolation is incorrect, an alternate solution would be to modify this class to create a temporary table in the replica (which is permitted in Aurora) and carry out all of the bulk inserts/updates, and then at the end of the `build_contact_rollups` method, iterate over that temp table, copying each row to the `contact_rollups_daily` table in the primary database.

File: lib/cdo/contact_rollups.rb [hunk 1]
```diff
@@ -682,7 +675,7 @@ def self.update_professional_learning_attendance_for_course_from_sections(course
 
   def self.mysql_multi_connection
     # return a connection with the MULTI_STATEMENTS flag set that allows multiple statements in one DB call
-    Sequel.connect(CDO.pegasus_reporting_db_writer.sub('mysql:', 'mysql2:'), flags: ::Mysql2::Client::MULTI_STATEMENTS)
+    Sequel.connect(CDO.PEGASUS_DB_WRITER.sub('mysql:', 'mysql2:'), flags: ::Mysql2::Client::MULTI_STATEMENTS)
```
Comment (wjordan) [inline lib/cdo/contact_rollups.rb]: This looks like a possible mistake from the find-and-replace?

**Comment by hacodeorg** [comment]
Do we have to update ContactRollupsValidation also since it still reads from PEGASUS_REPORTING_DB_READER

**Comment by sureshc** [comment]
The replica (reporting) database is typically only a few seconds behind the primary, so writes carried out in the primary database should be visible to the validation step.

File: lib/cdo/contact_rollups.rb (hunk 1 above)
Comment (sureshc) [reply lib/cdo/contact_rollups.rb]: Yes.  And would have been caught by the simple integration test that @uponthesun suggested to me this afternoon.  Fixing ...

**Comment by hacodeorg** [comment]
To make sure I understand, there are 4 main steps of contact rollups, 1) `build_contact_rollups`, 2) `validate_contact_rollups`, 3) `sync_contact_rollups_to_main` and 4) `sync_contact_rollups_to_pardot`. 3 of them (step 1, 3, 4) run in primary db and only step 2 runs in replica db?

File: lib/cdo/contact_rollups.rb (hunk 1 above)
Comment (sureshc) [reply lib/cdo/contact_rollups.rb]: fixed

**Comment by sureshc** [comment]
> 3 of them (step 1, 3, 4) run in primary db

Only step 1 is being changed here.  It is being modified so that all writes are to the primary database.  SELECT statements that feed into those writes are inherently also being modified to use the primary database, but other reads in step 1 will continue to run in the replica/reporting database

**Review by hacodeorg — APPROVED** [review]

**Comment by sureshc** [comment]
INF-129 is the JIRA we are here(-a) for.

**Pull request closed without merging**